### ISSUE TITLE
'try-except-raise' warnings on 3.9 but not 3.7

### ISSUE THREAD

**Issue by RemiCardona** [pylint-dev/pylint#5884]
### Bug description

```python
import asyncio

async def coro():
    try:
        await asyncio.sleep(0.5)
    except asyncio.CancelledError:
        raise
    except ValueError:
        pass
```

No warning raised when running in a python 3.7 venv, `W0706(try-except-raise)` warning raised with an identical pylint call when running in a 3.9 venv.

`# pylint: disable=try-except-raise,useless-suppression` thus needed to make pylint happy on both 3.7 and 3.9.

### Command used

```shell
pylint demo.py
```

### Pylint output

```shell
************* Module bluelink.demo
demo.py:7: [W0706(try-except-raise), coro] The except handler raises immediately
```

### Expected behavior

No warning raised, as on 3.7. It's a false positive.

### Pylint version

```shell
pylint 2.12.2
astroid 2.9.3
Python 3.7.12 (default, Mar  8 2022, 11:09:12) 
[GCC 11.2.1 20220115]

and

pylint 2.12.2
astroid 2.9.3
Python 3.9.10 (main, Mar  8 2022, 11:00:59) 
[GCC 11.2.1 20220115]
```

### OS / Environment

~amd64 Gentoo (outputs above) and Debian Buster (with different python patch-level versions)

**Comment by Pierre-Sassoulas** [comment]
Thank you for opening the issue. Are you using the ``py-version`` option (set to 3.7) ?

**Comment by RemiCardona** [comment]
Nope, I'm not using this option. And I can reproduce the bug with the provided sample and a plain call to `pylint demo.py`.

**Comment by RemiCardona** [comment]
I tried setting `py-version` to both `3.7` and `3.9`, under both venvs and the results are the same as before.

**Comment by jacobtylerwalls** [comment]
Thanks for writing. This is a false negative for `try-except-raise` in python 3.7. The warning is emitted correctly on Python 3.9.

Most likely, the failure to raise on Python 3.7 is because before Python 3.8 `asyncio.CancelledError` had a more complicated inheritance structure (it just aliased `concurrent.futures.CancelledError`). See also: https://github.com/PyCQA/pylint/issues/3083#issuecomment-841075906.

Ultimately, this is just a failure to infer `asyncio.CancelledError` in python 3.7:
```python
(Pdb) handler.type
<Attribute.CancelledError l.6 at 0x7fc4709ced90>
(Pdb) utils.safe_infer(handler.type)
Uninferable
```

Python 3.7 source, showing why it may be hard to infer this correctly:
https://github.com/python/cpython/blob/f656bc1cdbdfaaa07f66ed97e011b258b97e2788/Lib/asyncio/base_futures.py#L3-L9

I don't know that there's much we can do about this in `astroid`. I think you've found the best workaround.

**Comment by RemiCardona** [comment]
Why a false negative? I want to *not* handle that exception. I agree that
```python
try:
    stuff()
except:
    raise
``` definitely can and should be simplified, but in a situation with multiple except blocks, there's no other way to write the code to get the same behavior. It's not like an `if` statement that can be reversed.

**Comment by jacobtylerwalls** [comment]
I didn't write the checker, but I believe it's saying you can write:

```python
try:
    stuff()
# don't even bother catching asyncio.CancelledError
# that's fine, because it's not a subclass of ValueError
except ValueError:
    pass
```
Now, if it is possible for you to encounter other subclasses of _both_ CancelledError and ValueError, which is a diamond inheritance problem, then pylint's message is wrong and the code snippet is correct. I opened #5889 for that after reading your message. But I took the original report to be about the difference in behavior between versions and not about potential diamond inheritance problems from other unmentioned exception types.

**Comment by jacobtylerwalls** [comment]
Thanks again for the report.

Closing as the warning works as intended (and like many messages, if you prefer the verbosity of the immediate raise, you can disable the warning). We don't support Python 3.7 anymore so the difference before Python 3.8 is moot.
